// @lcu/common/esm5/lib/registry/enterprise/application.js
import { __extends } from "tslib";
var Application = /** @class */ (function () {
    function Application() {
    }
    return Application;
}());
export { Application };
var DAFApplication = /** @class */ (function () {
    function DAFApplication() {
    }
    return DAFApplication;
}());
export { DAFApplication };
var DAFAPIApplicationDetails = /** @class */ (function () {
    function DAFAPIApplicationDetails() {
    }
    return DAFAPIApplicationDetails;
}());
export { DAFAPIApplicationDetails };
var DAFViewApplicationDetails = /** @class */ (function () {
    function DAFViewApplicationDetails() {
    }
    return DAFViewApplicationDetails;
}());
export { DAFViewApplicationDetails };
var DAFLCUApplicationDetails = /** @class */ (function (_super) {
    __extends(DAFLCUApplicationDetails, _super);
    function DAFLCUApplicationDetails() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return DAFLCUApplicationDetails;
}(DAFViewApplicationDetails));
export { DAFLCUApplicationDetails };
var DAFRedirectApplicationDetails = /** @class */ (function () {
    function DAFRedirectApplicationDetails() {
    }
    return DAFRedirectApplicationDetails;
}());
export { DAFRedirectApplicationDetails };
var DAFAppPointerApplicationDetails = /** @class */ (function () {
    function DAFAppPointerApplicationDetails() {
    }
    return DAFAppPointerApplicationDetails;
}());
export { DAFAppPointerApplicationDetails };
export var DAFApplicationPackageTypes;
(function (DAFApplicationPackageTypes) {
    DAFApplicationPackageTypes["Git"] = "Git";
    DAFApplicationPackageTypes["NPM"] = "NPM";
    DAFApplicationPackageTypes["Zip"] = "Zip";
})(DAFApplicationPackageTypes || (DAFApplicationPackageTypes = {}));
var DAFApplicationGitPackage = /** @class */ (function () {
    function DAFApplicationGitPackage() {
    }
    return DAFApplicationGitPackage;
}());
export { DAFApplicationGitPackage };
var DAFApplicationNPMPackage = /** @class */ (function () {
    function DAFApplicationNPMPackage() {
    }
    return DAFApplicationNPMPackage;
}());
export { DAFApplicationNPMPackage };
var DAFApplicationZipPackage = /** @class */ (function () {
    function DAFApplicationZipPackage() {
    }
    return DAFApplicationZipPackage;
}());
export { DAFApplicationZipPackage };
