// @lcu/common/esm2015/lib/modules/pipe.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DataPipes } from './../utils/pipes/data.pipes';
import { SanitizeHtmlPipe } from './../utils/pipes/sanitze-html.pipe';
import { DecamelizePipe } from '../utils/pipes/decamelize.pipe';
import { StringPipes } from '../utils/pipes/string.pipes';
export class PipeModule {
    static forRoot() {
        return {
            ngModule: PipeModule,
            providers: [],
        };
    }
}
PipeModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    DataPipes,
                    SanitizeHtmlPipe,
                    DecamelizePipe,
                    StringPipes
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    DataPipes,
                    SanitizeHtmlPipe,
                    DecamelizePipe,
                    StringPipes
                ]
            },] }
];
