// @lcu/common/esm5/lib/modules/pipe.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DataPipes } from './../utils/pipes/data.pipes';
import { SanitizeHtmlPipe } from './../utils/pipes/sanitze-html.pipe';
import { DecamelizePipe } from '../utils/pipes/decamelize.pipe';
import { StringPipes } from '../utils/pipes/string.pipes';
var PipeModule = /** @class */ (function () {
    function PipeModule() {
    }
    PipeModule.forRoot = function () {
        return {
            ngModule: PipeModule,
            providers: [],
        };
    };
    PipeModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DataPipes,
                        SanitizeHtmlPipe,
                        DecamelizePipe,
                        StringPipes
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        DataPipes,
                        SanitizeHtmlPipe,
                        DecamelizePipe,
                        StringPipes
                    ]
                },] }
    ];
    return PipeModule;
}());
export { PipeModule };
